File: leetcode/C++/odd-even-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        ListNode *odd=NULL, *even=NULL, *cur=head, *es=NULL;
        for(int i=0; cur; cur=cur->next,++i){
            if(i%2==0) {
                if(!odd) odd=cur;
                else {
                    odd->next=cur;
                    odd=odd->next;
                }
            }
            else {
                if(!even) es=even=cur;
                else{
                    even->next=cur;
                    even=even->next;
                }
            }
        }
        if(odd) odd->next=es;
        if(even) even->next=NULL;
        return head;
    }
};

File: leetcode/Python/odd-even-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: ListNode) -> ListNode:
        odd, even, es, cur, i = None, None, None, head, 0
        while cur:
            if i%2:
                if not even:
                    es=even=cur
                else:
                    even.next=cur
                    even=even.next
            else:
                if not odd:
                    odd=cur
                else:
                    odd.next=cur
                    odd=odd.next
            cur=cur.next
            i+=1
        if odd:
            odd.next=es
        if even:
            even.next=None
        return head